File: src/de/kaffeeshare/server/plugins/XKCD.java
package de.kaffeeshare.server.plugins;

import java.net.URL;

import org.jsoup.nodes.Document;

import de.kaffeeshare.server.exception.PluginErrorException;

public class XKCD extends BasePlugin {

	@Override
	public boolean match(URL url) {
		return (match(url, "xkcd.com") || match(url, "www.xkcd.com"));
	}
	
	@Override
	protected String getDescription(Document doc) {

		String description = null;

		try {
			description  = doc.getElementById("comic").html();
		} catch (Exception e) {
			throw new PluginErrorException(this);
		}
		
		return description;
	}

	@Override
	protected String getCaption(Document doc) {
		return "XKCD - " + doc.getElementById("ctitle").html();
	}
	
	@Override
	protected String getImageUrl(Document doc) {
		return null;
	}
}

File: extract/plugins/xkcd.go
package plugins

import (
	"bytes"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"github.com/koffeinsource/kaffeeshare/data"

	"golang.org/x/net/context"
	"golang.org/x/net/html"
	"google.golang.org/appengine/log"
)

// Xkcd extract the comic from an XKCD page
func Xkcd(i *data.Item, sourceURL string, doc *goquery.Document, c context.Context) {
	if !strings.Contains(sourceURL, "xkcd.com") {
		return
	}

	log.Infof(c, "Running XKCD plugin.")

	selection := doc.Find("#comic")
	if len(selection.Nodes) == 0 {
		log.Infof(c, "XKCD plugin found no #comic. "+sourceURL)
	} else {
		if len(selection.Nodes) > 1 {
			log.Infof(c, "XKCD plugin found >1 #comic. "+sourceURL)
		}
		buf := new(bytes.Buffer)
		err := html.Render(buf, selection.Nodes[0])
		if err != nil {
			log.Infof(c, "XKCD plugin error while rendering. "+sourceURL+"- "+err.Error())
			return
		}
		i.Description = buf.String()
	}

	selection = doc.Find("#ctitle")
	if len(selection.Nodes) == 0 {
		log.Infof(c, "XKCD plugin found no #ctitle. "+sourceURL)
	} else {
		if len(selection.Nodes) > 1 {
			log.Infof(c, "XKCD plugin found >1 #ctitle. "+sourceURL)
		}
		if selection.Nodes[0].FirstChild != nil {
			i.Caption = "XKCD - " + selection.Nodes[0].FirstChild.Data
		} else {
			i.Caption = "XKCD"
		}
	}

	i.ImageURL = ""

}
